libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs

```diff
@@ -151,6 +159,52 @@ instance DecCBOR RDPair where
   decCBOR =
     decodeRecordNamed "RDPair" (const 2) $ RDPair <$> decCBOR <*> decCBOR
 
+data DRep c
+  = DRepKeyHash (KeyHash 'Voting c)
+  | DRepScriptHash (ScriptHash c)
+  | DRepAlwaysAbstain
+  | DRepAlwaysNoConfidence
+  deriving (Show, Eq, Ord, Generic, NoThunks, NFData, ToExpr, ToJSON)
+
+instance Crypto c => EncCBOR (DRep c) where
+  encCBOR (DRepKeyHash kh) =
+    encode $
+      Sum DRepKeyHash 0
+        !> To kh
+  encCBOR (DRepScriptHash sh) =
+    encode $
+      Sum DRepScriptHash 1
+        !> To sh
+  encCBOR DRepAlwaysAbstain =
+    encode $
+      Sum DRepAlwaysAbstain 2
+  encCBOR DRepAlwaysNoConfidence =
+    encode $
+      Sum DRepAlwaysNoConfidence 3
+
+instance Crypto c => DecCBOR (DRep c) where
+  decCBOR = decode $
+    Summands "DRep" $ \case
+      0 -> SumD DRepKeyHash <! From
+      1 -> SumD DRepScriptHash <! From
+      2 -> SumD DRepAlwaysAbstain
+      3 -> SumD DRepAlwaysNoConfidence
+      k -> Invalid k
+
+dRepToCred :: DRep c -> Maybe (Credential 'Voting c)
+dRepToCred (DRepKeyHash kh) = Just $ KeyHashObj kh
+dRepToCred (DRepScriptHash sh) = Just $ ScriptHashObj sh
+dRepToCred _ = Nothing
+
+pattern DRepCredential :: Credential 'Voting c -> DRep c
+pattern DRepCredential c <- (dRepToCred -> Just c)
+  where
+    DRepCredential c = case c of
+      ScriptHashObj sh -> DRepScriptHash sh
+      KeyHashObj kh -> DRepKeyHash kh
+
+{-# COMPLETE DRepCredential, DRepAlwaysAbstain, DRepAlwaysNoConfidence :: DRep #-}
```

Comment:
Although I agree that `UMap` isn't completely agnostic to our particular domain, I think that may be the `DRep` data type is more suited to reside along with `Credential` in its module?

Reply:
I feel like the `UMap` module is more fitting, because it has to do with governance and it only appears in the `UMap`, whereas Credentials are used everywhere.